File: services/cookbook/core/src/main/java/family/haschka/wolkenschloss/cookbook/recipe/Rational.java
package family.haschka.wolkenschloss.cookbook.recipe;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public record Rational(int numerator, int denominator)  {

    public Rational {
        if (denominator == 0) {
            throw new InvalidNumber();
        }

        var gcd = gcd(numerator, denominator);
        numerator /= gcd;
        denominator /= gcd;
    }

    public Rational(int number, int numerator, int denominator) {
        this(numerator + number * denominator, denominator);
    }

    private static final Map<Rational, Character> codes;
    private static final Map<String, Rational> fractions;
    public static final String REGEX;
    private static final Pattern pattern;

    static {
        codes = new HashMap<>();
        codes.put(new Rational(1, 2), '\u00bd');
        codes.put(new Rational(1, 3), '\u2153');
        codes.put(new Rational(2, 3), '\u2154');
        codes.put(new Rational(1, 4), '\u00bc');
        codes.put(new Rational(3, 4), '\u00be');
        codes.put(new Rational(1, 5), '\u2155');
        codes.put(new Rational(2, 5), '\u2156');
        codes.put(new Rational(3, 5), '\u2157');
        codes.put(new Rational(4, 5), '\u2158');
        codes.put(new Rational(1, 6), '\u2159');
        codes.put(new Rational(5, 6), '\u215a');
        codes.put(new Rational(1, 7), '\u2150');
        codes.put(new Rational(1, 8), '\u215b');
        codes.put(new Rational(3, 8), '\u215c');
        codes.put(new Rational(5, 8), '\u215d');
        codes.put(new Rational(7, 8), '\u215e');
        codes.put(new Rational(1, 9), '\u2151');
        codes.put(new Rational(1, 10), '\u2152');

        REGEX = "((?<number>-?(0|[1-9]\\d*)(?<!-0))[ ]?)?(((?<numerator>-?(0|[1-9]\\d*)(?<!-0))([/](?<denominator>[1-9]\\d*))*)|(?<fraction>[½⅔¾⅘⅚⅞⅓⅗¼⅖⅝⅕⅙⅜⅐⅛⅑⅒]))?";
        pattern = Pattern.compile("^" + REGEX + "$");

        fractions = new HashMap<>();
        for (Rational r : codes.keySet()) {
            fractions.put(codes.get(r).toString(), r);
        }
    }

    public Rational(int value) {
        this(value, 1);
    }

    private int gcd(int a, int b) {
        if (b == 0) return a;
        return gcd(b, a % b);
    }

    public Rational multiply(Rational b) {
        return new Rational(numerator * b.numerator, denominator * b.denominator);
    }

    public static Rational parse(String input) {

        Matcher m = pattern.matcher(input);
        if (m.find()) {
            String number = m.group("number");
            String fraction = m.group("fraction");
            String numerator = m.group("numerator");
            String denominator = m.group("denominator");

            if (number != null) {
                if (fraction != null) {
                    return fractions.get(fraction)
                            .add(new Rational(Integer.parseInt(number), 1));
                } else if (numerator != null) {
                    return new Rational(
                            Integer.parseInt(number),
                            Integer.parseInt(numerator),
                            Integer.parseInt(denominator));
                } else {
                    return new Rational(Integer.parseInt(number));
                }
            } else {
                if (fraction != null) {
                    return fractions.get(fraction);
                } else if (numerator != null) {
                    return new Rational(Integer.parseInt(numerator), Integer.parseInt(denominator));
                }
            }
        }

        throw new InvalidNumber();
    }

    public String toString() {
        StringBuilder builder = new StringBuilder();

        if (denominator == 1) {
            builder.append(numerator);
        } else {
            var number = numerator / denominator;

            if (number != 0) {
                builder.append(number);
                builder.append(" ");

                var rest = this.minus(new Rational(number, 1));
                appendRational(builder, rest);
            } else {
                appendRational(builder, this);
            }
        }

        return builder.toString();
    }

    private void appendRational(StringBuilder builder, Rational rest) {
        if (codes.containsKey(rest)) {
            builder.append(codes.get(rest));
        } else {
            builder.append(rest.numerator);
            builder.append("/");
            builder.append(rest.denominator);
        }
    }

    public Rational minus(Rational rational) {
        var d = denominator * rational.denominator;
        var a = numerator * rational.denominator;
        var b = rational.numerator * denominator;
        return new Rational(a - b, d);
    }

    public Rational add(Rational rational) {
        var d = denominator * rational.denominator;
        var a = numerator * rational.denominator;
        var b = rational.numerator * denominator;

        return new Rational(a + b, d);
    }
}

File: services/cookbook/core/src/main/kotlin/recipe/Rational.kt
package family.haschka.wolkenschloss.cookbook.recipe

import family.haschka.wolkenschloss.cookbook.parser.IngredientLexer
import family.haschka.wolkenschloss.cookbook.parser.IngredientParser
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream
import kotlin.math.absoluteValue

data class Rational(private var numerator: Int, private var denominator: Int) {

    constructor(value: Int) : this(value, 1)

    init {
        if (denominator == 0) {
            throw InvalidNumber("Denominator must not be 0")
        }

        val gcd = gcd(numerator, denominator)

        numerator /= gcd
        denominator /= gcd
    }

    private fun gcd(a: Int, b: Int): Int {
        return if (b == 0) a else gcd(b, a % b)
    }


    override fun toString(): String {
        val builder = StringBuilder()
        if (denominator == 1) {
            builder.append(numerator)
        } else {
            if ((numerator * denominator) < 0) {
                builder.append('-')
            }
            val number: Int = numerator.absoluteValue / denominator.absoluteValue
            if (number != 0) {
                builder.append(number)
                builder.append(" ")
                val rest = this.absoluteValue().minus(Rational(number, 1))
                appendRational(builder, rest)
            } else {
                appendRational(builder, this.absoluteValue())
            }
        }
        return builder.toString()
    }

    private fun appendRational(builder: StringBuilder, rest: Rational) {
        if (codes.containsKey(rest)) {
            builder.append(codes[rest])
        } else {
            builder.append(rest.numerator)
            builder.append("/")
            builder.append(rest.denominator)
        }
    }

    operator fun minus(rational: Rational): Rational {
        val d: Int = denominator * rational.denominator
        val a: Int = numerator * rational.denominator
        val b: Int = rational.numerator * denominator
        return Rational(a - b, d)
    }

    operator fun plus(rational: Rational): Rational {
        val d: Int = denominator * rational.denominator
        val a: Int = numerator * rational.denominator
        val b: Int = rational.numerator * denominator
        return Rational(a + b, d)
    }

    operator fun times(b: Rational): Rational {
        return Rational(numerator * b.numerator, denominator * b.denominator)
    }

    operator fun div(b: Rational): Rational {
        return this.times(Rational(b.denominator, b.numerator))
    }

    fun absoluteValue(): Rational {
        return Rational(this.numerator.absoluteValue, this.denominator.absoluteValue)
    }

    companion object {
        val codes: Map<Rational, Char> = hashMapOf(
            Rational(1, 2) to '\u00bd',
            Rational(1, 3) to '\u2153',
            Rational(2, 3) to '\u2154',
            Rational(1, 4) to '\u00bc',
            Rational(3, 4) to '\u00be',
            Rational(1, 5) to '\u2155',
            Rational(2, 5) to '\u2156',
            Rational(3, 5) to '\u2157',
            Rational(4, 5) to '\u2158',
            Rational(1, 6) to '\u2159',
            Rational(5, 6) to '\u215a',
            Rational(1, 7) to '\u2150',
            Rational(1, 8) to '\u215b',
            Rational(3, 8) to '\u215c',
            Rational(5, 8) to '\u215d',
            Rational(7, 8) to '\u215e',
            Rational(1, 9) to '\u2151',
            Rational(1, 10) to '\u2152',
        )

        val fractions: Map<String, Rational> =
            codes.entries.associate { (key, value) -> value.toString() to key }

        fun parse(input: String): Rational {
            val stream = CharStreams.fromString(input)
            val lexer = IngredientLexer(stream)
            val tokens = CommonTokenStream(lexer)
            val parser = IngredientParser(tokens)
            val listener = SyntaxErrorListener()
            parser.addErrorListener(listener)
            val tree = parser.mixed_fraction()

            if (parser.numberOfSyntaxErrors > 0) {
                throw InvalidNumber(listener.errors.joinToString(", "))
            }

            return RationalVisitor().visit(tree)
        }
    }
}